### PR TITLE
Fixed the welcome message contains single quote character

### LINKED ISSUE #35481: [Issue] Fixed the welcome message contains single quote character
This issue is automatically created based on existing pull request: magento/magento2#34061: Fixed the welcome message contains single quote character

---------

### Description (*)

This PR is fixed for the issue, that if we put the single quote character to the header welcome message config, it will lead the frontend to have a bunch of console errors:

> Uncaught SyntaxError: Unable to process binding "ifnot: function(){return customer().fullname }"
Message: Unable to parse bindings.

### Manual testing scenarios (*)

1. From Admin, navigate to Content -> Design -> Configuration
2. Select a theme and Edit
3. In the edit page, expand the Header setting, put the `Welcome Text`, like "Welcome Khoi's store!"
4. Clear cache, then go to frontend, you'll see a JS error, and some JS functions not working because of this issue.

### Questions or comments
Maybe we need to bind the message as a text, instead of HTML binding, like:

> data-bind="text: '<?= $escaper->escapeHtmlAttr($welcomeMessage) ?>'"

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] README.md files for modified modules are updated and included in the pull request if any [README.md predefined sections](https://github.com/magento/devdocs/wiki/Magento-module-README.md) require an update
 - [ ] All automated tests passed successfully (all builds are green)

### PR DESCRIPTION
**Pull request by khoimm92**
### Description (*)

This PR is fixed for the issue, that if we put the single quote character to the header welcome message config, it will lead the frontend to have a bunch of console errors:

> Uncaught SyntaxError: Unable to process binding "ifnot: function(){return customer().fullname }"
Message: Unable to parse bindings.

### Manual testing scenarios (*)

1. From Admin, navigate to Content -> Design -> Configuration
2. Select a theme and Edit
3. In the edit page, expand the Header setting, put the `Welcome Text`, like "Welcome Khoi's store!"
4. Clear cache, then go to frontend, you'll see a JS error, and some JS functions not working because of this issue.

### Questions or comments
Maybe we need to bind the message as a text, instead of HTML binding, like:

> data-bind="text: '<?= $escaper->escapeHtmlAttr($welcomeMessage) ?>'"

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] README.md files for modified modules are updated and included in the pull request if any [README.md predefined sections](https://github.com/magento/devdocs/wiki/Magento-module-README.md) require an update
 - [ ] All automated tests passed successfully (all builds are green)

### Resolved issues:
1. [x] resolves magento/magento2#35481: Fixed the welcome message contains single quote character

### CODE REVIEW

**Comment by khoimm92** [comment]
@magento run all tests

**Comment by mrtuvn** [comment]
imho should be text in that point

**Comment by BarnyShergold** [comment]
@magento run all tests

**Comment by BarnyShergold** [comment]
@khoimm92 - You need to sign the Adobe CLA as part of this PR review - you only ever need to do this once.

**Comment by khoimm92** [comment]
@magento run all tests

**Comment by BarnyShergold** [comment]
@magento run Static Tests

**Comment by BarnyShergold** [comment]
@khoimm92 - You will need to fix the Static test on this - not sure why they passed before!

**Comment by BarnyShergold** [comment]
@khoimm92 - You also need to sign the Adobe CLA

**Comment by BarnyShergold** [comment]
@magento run all tests

**Review by Den4ik — CHANGES_REQUESTED** [review]
@khoimm92 Please fix failed static tests

**Comment by mrtuvn** [comment]
wt
html not allowed

**Comment by khoimm92** [comment]
Thanks @mrtuvn, let me change that.
@BarnyShergold @Den4ik I got it. Also, my fix did not cover the case putting the double quote. Let me have a tweak for this.

**Comment by engcom-Bravo** [comment]
@magento run all tests

**Comment by engcom-Bravo** [comment]
@magento run all tests

**Comment by engcom-Bravo** [comment]
@magento run all tests

**Comment by engcom-Bravo** [comment]
@magento run Functional Tests CE,Functional Tests EE,Functional Tests B2B

**Comment by engcom-Bravo** [comment]
@magento run all tests

**Comment by engcom-Bravo** [comment]
@magento run Functional Tests CE,Functional Tests EE,Functional Tests B2B

**Comment by engcom-Bravo** [comment]
@magento run all tests

**Review by Den4ik — CHANGES_REQUESTED** [review]
@khoimm92 Please fix failed Static Tests

**Comment by engcom-Bravo** [comment]
@magento run all tests

**Comment by engcom-Bravo** [comment]
@magento run Functional Tests CE,Functional Tests B2B

**Review by ishakhsuvarov — APPROVED** [review]

**Comment by engcom-Alfa** [comment]
@magento run all tests

**Comment by engcom-Alfa** [comment]
@magento Functional Tests B2B, Functional Tests CE, Functional Tests EE

**Comment by engcom-Alfa** [comment]
@magento run Functional Tests B2B, Functional Tests CE, Functional Tests EE

**Comment by engcom-Alfa** [comment]
@magento run Functional Tests B2B, Functional Tests CE, Functional Tests EE

**Comment by engcom-Alfa** [comment]
https://github.com/magento run Functional Tests B2B, Functional Tests EE

**Comment by engcom-Alfa** [comment]
@magento run Functional Tests B2B, Functional Tests EE

**Comment by engcom-Alfa** [comment]
:heavy_check_mark: QA Passed

**Preconditions:**

1. Have Magento 2.4-develop latest code installed


**Manual testing scenario:**

1. Admin login> Content> Configuration> Header section> Edit the Welcome Message to something like `"Welcome XYZ"` (Here, double quotes is a part of message). Save the configurations.

2. Open the front end customer side, and valid the message JS errors in Console as well.

**Before:** :heavy_multiplication_x: There was a JS Syntax error in the console and as well as message was not displayed.
image

**After:** :heavy_check_mark: Message is displayed accordingly and JS Syntax error is not existing anymore.
image

It is a specific case affecting functional and home presentation, will be added with the regression test case covering a few more input types as well.

**Comment by engcom-Alfa** [comment]
Since a couple of jobs are getting failed continuously even after multiple tries, moving this PR ticket to Extended Testing column for further investigation.

**Comment by engcom-Charlie** [comment]
@magento run Functional Tests B2B, Functional Tests EE

**Comment by engcom-Charlie** [comment]
It seems bot moved this PR from Extended Testing to Ready for testing after assignment, so moved it back to Extended Testing.

**Comment by engcom-Charlie** [comment]
Since the build is green now and as per this above [comment](https://github.com/magento/magento2/pull/34061#issuecomment-1110931454) PR is QA passed so moving it to Merge in Progress.

Thank you!

**Comment by Den4ik** [comment]
@engcom-Charlie  did you test it with another characters that require slash. For example, double quotes?

File: app/code/Magento/Theme/Block/Html/Header.php [hunk 1]
```diff
@@ -34,6 +34,7 @@ public function getWelcome()
                 \Magento\Store\Model\ScopeInterface::SCOPE_STORE
             );
         }
+        $this->_data['welcome'] = str_replace("'", "\'", $this->_data['welcome']);
```
Comment (Den4ik) [inline app/code/Magento/Theme/Block/Html/Header.php]: I believe that use off `addslashes` is more preferable here

**Review by Den4ik — CHANGES_REQUESTED** [review]
Hey @khoimm92 
Could you check my comment

**Comment by engcom-Charlie** [comment]
> @engcom-Charlie did you test it with another characters that require slash. For example, double quotes?

Hi @Den4ik,

Sure, will look into your review comments.

Thank you!

**Comment by engcom-Charlie** [comment]
Hi @Den4ik,

I have mode the modification suggested by you. addslashes() is working well for all combinations of welcome message. Committed and pushed the code.

Moving this PR to Ready for Testing now.

Thank you!

**Comment by engcom-Charlie** [comment]
@magento run all tests

**Comment by engcom-Lima** [comment]
@magento run all tests

**Comment by engcom-Lima** [comment]
@magento run Integration Tests Static Tests

**Comment by engcom-Lima** [comment]
@magento run Integration Tests,Static Tests

**Comment by engcom-Lima** [comment]
@magento run Integration Tests,Static Tests

**Comment by engcom-Lima** [comment]
:heavy_check_mark: QA Passed

**Preconditions:**

1. Install the fresh Magento 2.4-develop instance.
2. Have a few products already created in the instance

**Manual testing scenario:**

1. Login to Admin, navigate to Content -> Design -> Configuration
2. Select a theme and Edit
3. In the edit page, expand the Header setting, put the Welcome Text, like "Welcome to our's store!"
4. Clear cache, then go to frontend and open the console

**Before:** :heavy_multiplication_x: Getting JS error, and some JS functions not working because of this issue.

jserror

**After:** :heavy_check_mark:  There was no js error found in the console after PR changes.
Screenshot from 2022-05-13 16-39-03

Tried to use different type of symbols and validated if any error visible in the front-end.

**Comment by engcom-Charlie** [comment]
@magento run Integration Tests, Static Tests

**Comment by engcom-Charlie** [comment]
After making [these](https://github.com/magento/magento2/pull/34061#issuecomment-1124906069) changes was getting below static test failure. 

image

Implemented the fix for the same and pushed the code base.

Thank you!

**Comment by engcom-Charlie** [comment]
@magento run all tests

**Comment by engcom-Lima** [comment]
@magento run all tests

**Comment by engcom-Lima** [comment]
@magento run Unit Tests, Static Tests, Semantic Version Checker

**Comment by engcom-Lima** [comment]
@magento run Static Tests

**Comment by engcom-Lima** [comment]
@magento give me test instance

**Comment by engcom-Charlie** [comment]
@magento run Static Tests, Semantic Version Checker, Unit Tests

**Comment by engcom-Charlie** [comment]
@magento run all tests

File: app/code/Magento/Theme/Block/Html/Header.php [hunk 2]
```diff
@@ -14,6 +16,27 @@
...
+     * @param \Magento\Framework\View\Element\Template\Context $context
+     * @param Magento\Framework\Escaper $escaper
+     * @param array $data
+     */
+    public function __construct(
+        \Magento\Framework\View\Element\Template\Context $context,
+        \Magento\Framework\Escaper $escaper,
+        array $data = []
+    ) {
+        $this->_escaper = $escaper;
```
Comment (Den4ik) [inline app/code/Magento/Theme/Block/Html/Header.php]: Property declared at line 22 don't has underscore.
Please don't use underscored variables and properties.

**Review by Den4ik — CHANGES_REQUESTED** [review]
Hi @engcom-Charlie 
Thanks for your update. Please look at my comment

**Comment by engcom-Charlie** [comment]
Hi @Den4ik,

Thank you for reviewing the changes.

I have modified the implementation as per your suggested changes and pushed the code.

Thank you!

**Comment by engcom-Charlie** [comment]
@magento run all tests

**Comment by engcom-Charlie** [comment]
@magento run Functional Tests B2B, Functional Tests CE, Functional Tests EE, Semantic Version Checker, Unit Tests

**Comment by engcom-Charlie** [comment]
Fixed the build failure. Implemented the fix for unit test and pushed the code for the same.

Moving it to Testing now.

Thank you!

**Comment by engcom-Alfa** [comment]
@magento run Functional Tests EE, Integration Tests, Semantic Version Checker

**Comment by engcom-Lima** [comment]
@magento run Semantic Version Checker

**Comment by engcom-Lima** [comment]
@magento run Semantic Version Checker

**Comment by engcom-Lima** [comment]
Tested this PR, working fine but Semantic Version Checker is failed. Hence, raising [JIRA](https://jira.corp.magento.com/browse/MC-43367) approval for svc failure.

**Comment by engcom-Charlie** [comment]
It seems that bot moved the PR from Pending Approval to Changes Requested. Moving it back.

Thank you!

**Pull request merged**